Dwolla.Client/HttpServices/Architecture/BaseHttpService.cs

```diff
@@ -0,0 +1,80 @@
+﻿using Dwolla.Client.Models;
+using Dwolla.Client.Models.Responses;
+using Dwolla.Client.Rest;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Dwolla.Client.HttpServices.Architecture
+{
+	public abstract class BaseHttpService
+	{
+		internal Task<string> getAccessToken;
+		internal IDwollaClient client;
+
+		public BaseHttpService(IDwollaClient dwollaClient, Task<string> getAccessToken)
+		{
+			client = dwollaClient;
+			this.getAccessToken = getAccessToken;
+		}
+
+		internal async Task<RestResponse<TRes>> GetAsync<TRes>(Uri uri, CancellationToken cancellationToken = default) where TRes : IDwollaResponse
+		{
+			var headers = await CreateHeaders();
+
+			return await ExecAsync(() => client.GetAsync<TRes>(uri, headers, cancellationToken));
+		}
+
+		internal async Task<RestResponse<EmptyResponse>> PostAsync<TReq>(Uri uri, TReq request, CancellationToken cancellationToken = default)
+		{
+			var headers = await CreateHeaders();
+
+			return await ExecAsync(() => client.PostAsync<TReq, EmptyResponse>(uri, request, headers, cancellationToken));
+		}
+
+		internal async Task<RestResponse<TRes>> PostAsync<TReq, TRes>(Uri uri, TReq request, string idempotencyKey, CancellationToken cancellationToken = default) where TRes : IDwollaResponse
+		{
+			var headers = await CreateHeaders(idempotencyKey);
+
+			return await ExecAsync(() => client.PostAsync<TReq, TRes>(uri, request, headers, cancellationToken));
+		}
+
+		internal async Task<RestResponse<EmptyResponse>> DeleteAsync<TReq>(Uri uri, TReq request)
+		{
+			var headers = await CreateHeaders();
+
+			return await ExecAsync(() => client.DeleteAsync(uri, request, headers));
+		}
+
+		private async Task<Headers> CreateHeaders(string idempotencyKey = null)
+		{
+			var headers = new Headers
+			{
+				["Authorization"] = $"Bearer {await getAccessToken}"
```

Comment:
I think we should cache the expiration time along with the token. If we know the token is expired there's no reason to waste a request that we know will fail